RFCs/FS-1060-nullable-reference-types.md

```diff
@@ -0,0 +1,1126 @@
+
+As descibed in the [Interaction Model](FS-1060-nullable-reference-types.md#interaction-model), there are certain behaviors involved in how to work with assemblies that may or may not express nullability. The most flexible system for this is one in which a given scope is defined to have nullability as a concept. This is quite granular, but in practice will be at the assembly module level, making it mostly as if it were a per-project configuration.
+
+C# 8.0 code will emit the `[NonNullTypes(true|false)]` attribute:
+
+```csharp
+namespace System.Runtime.CompilerServices
+{
+    [AttributeUsage(AttributeTargets.Class |
+                    AttributeTargets.Constructor |
+                    AttributeTargets.Delegate |
+                    AttributeTargets.Enum |
+                    AttributeTargets.Event |
+                    AttributeTargets.Field |
+                    AttributeTargets.Interface |
+                    AttributeTargets.Method |
+                    AttributeTargets.Module |
+                    AttributeTargets.Property |
+                    AttributeTargets.Struct,
+                    AllowMultiple = false)]
+    public sealed class NonNullTypesAttribute : Attribute
+    {
+        public NonNullTypesAttribute(bool enabled = true) { }
+    }
+}
+```
+From the C# spec on this:
+
+> Unannotated reference types are non-nullable or [null-oblivious](FS-1060-nullable-reference-types.md#nullability-obliviousness)) depending on whether the containing scope includes `[NonNullTypes(true|false)]`.
+>
+> `[NonNullTypes(true|false)]` is not synthesized by the compiler. If the attribute is used explicitly in source, the type declaration must be provided explicitly to the compilation.
+
+In other words, this attribute specifies a way to mark a class, constructor, delegate, enum, event, field, interface, method, module, property, or struct as having nullability expressable or not in a containing scope.
+
+This attribute could potentially be used in F# to allow for opt-in/opt-out nullability at a fine-grained level.
+
+F# will respect this attribute, with the `true` caseindicating that the scope distinguishes between nullable and non-nullable reference types.
+
+F# will treat scopes with the `false` case as if its contained reference types are all nullable reference types.
+
+In other words, if we cannot be sure that a given reference type is non-nullable, we assume it is nullable.
+
+As a final note, this attribute _can_ be abused. For example, if a method is annotated with `[NonNullTypes(true)]` and does not actually perform a `null` check, further code can still produce a `NullReferenceException` and not have a warning associated with it.
+
+#### Null obliviousness
+
+Nullability obliviousness is a concept that C# 8.0 has. A null-oblivious type is one that no assumptions can be made about. Once assigned to a nullable or non-nullable variable, it is treated as if it is nullable or non-nullable, respectively.
```

Comment:
I need more information on this - e.g. examples where would C# infer null-obliviousness.

The natural corresponding thing in F# would be to have nullability inference variables.  For example, the most natural implementation in F# for nullability inference would mean that even the literal "a" would initially be assigned type `string | ?` where the `?` represents an inference variable which can be filled in with values "0" (indicating `string`) or "1" (indicating `string | null`).  Thus "obliviousness" is just a dangling inference variable.

This approach is used partly because we never merge to a "common super type" during type inference.  So when we have

    let f (s: string | null)  = if b then "a" else s

the `then` and `else` branches need the same type.  The type of `s` is `string | null` and the type of the literal `"a"` needs to be compatible with this.

Normally this approach to annotation inference is sufficient for F# (e.g. for units of measure and type inference) and means we won't need extra concepts like "null obliviousness" - they should just drop out naturally.  A good corresponding example is that `1.0<_>` is "unit oblivious")

Reply:
> Reference types we cannot determine to be non-nullable will be assumed to be nullable.**

I'm not sure where "assumed to be nullable" is coming dfrom or why we would do that for F#.  The only things we would assume to be nullable are things coming from .NET assemblies, and even that should be explicit assuming the assembly is annotated.

Note this would still apply even if we use nullability inference.  Nullability inference variables that are not "filled in" by the end of inference would be given a default value.  I believe in all situations for regular F# code that default value would be "non-nullable".

If there are counter examples we should add them

Reply:
The main gist of this (which I'll clarify):

* If C# does not see `NonNullTypes(false)` for a scope, it treats that scope as null-oblivious
* If something cannot be inferred to be nullable or non-nullable (`NonNullTypes(false)`), then **nullabulity or non-nullabulity is not inferred**
* Once a value is assigned to a variable, it is treated as either nullable or non-nullable (e.g., assigning to `string?` or `string`), based on how it is assigned
* The variable is treated as nullable or non-nullable from that point forward
* `var x = ...` is still undecided, since type inference means that you need to make a decision one way or another

Reply:
From the F# perspective, that `var x = ...` case is much more of a problem, since we use type inference everywhere. So I'm leaning towards us not having nullability obliviousness as a concept that we surface in user code.